pkg/server/generic_server.go

```diff
@@ -0,0 +1,107 @@
+package server
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+
+	arkcontroller "github.com/heptio/ark/pkg/controller"
+)
+
+type Interface interface {
+	Run() error
+	AddController(controller arkcontroller.Interface)
+}
+
+type genericServer struct {
+	logger               logrus.FieldLogger
+	controllers          []arkcontroller.Interface
+	preStartControllers  func(ctx context.Context, cancelFunc context.CancelFunc) error
+	postStartControllers func(ctx context.Context, cancelFunc context.CancelFunc) error
+	cleanup              func()
+	ctx                  context.Context
+	cancelFunc           context.CancelFunc
+}
+
+func NewGenericServer(
```

Comment:
I am not sure this is helpful. There is not enough commonality between the daemonset command and the server command.

DS:
- ✅ sets `controllers`
- ❌ doesn't set `preStartControllers`
- ✅ sets `postStartControllers`
- ❌ doesn't set `cleanup`

Server:
- ❌ doesn't set `controllers`
- ✅ sets `preStartControllers`
- ✅ sets `postStartControllers`
- ✅ sets `cleanup`

If we decide to keep this, it would look better using functional options, so the constructor would be reduced to
```go
func NewGenericServer(logger logrus.FieldLogger, options ...Option)
```

The common functionality used by both DS and Server is essentially:
- handling shutdown signals
- running controllers (although the registration is handled differently between DS and Server)

Pre/PostStartControllers and Cleanup are not used by both.

I don't think it's a burden to have 2 independent `Run()` implementations each invoke `handleShutdownSignals`.

We could consider creating a "controller executor" like this:
```go
type ControllerRegistration struct {
  controller controller.Interface
  workers int
}

type ControllerExecutor struct {
  controllers []ControllerRegistration
}

func (ce *ControllerExecutor) Execute(ctx context.Context) {
        var wg sync.WaitGroup
	for _, c := range ce.controllers {
		wg.Add(1)
		go func(c ControllerRegistration) {
			c.controller.Run(ctx, c.workers)
                        wg.Done()
		}(c)
	}
}
```

Both the DS and Server structs could use this. I think I'd prefer an approach in this vein as opposed to the genericServer.

Reply:
OK. I'll back this out.